source/common/http/conn_manager_impl.cc

```diff
@@ -2048,6 +2048,17 @@ void ConnectionManagerImpl::ActiveStream::encodeTrailers(ActiveStreamEncoderFilt
     }
   }
 
+  // Some browsers (e.g. WebKit-based browsers: https://bugs.webkit.org/show_bug.cgi?id=210108) have
+  // a problem with processing empty trailers of an HTTP/2 response as reported here:
+  // https://github.com/envoyproxy/envoy/issues/10514. This runtime feature flag is disabled by
+  // default.
+  const bool skip_encoding_empty_trailers =
+      Runtime::runtimeFeatureEnabled("envoy.reloadable_features.skip_encoding_empty_trailers");
```

Comment:
Thanks for un-breaking Safari, @dio  :-)  

If this is a common client-side bug, I'd think we want to make sure it "just works" by default.  

If this sort of transformation is legal by default (like taking an empty data frame and sending FIN in headers, but I'm not sure about the H2 spec here.  @PiotrSikora ?) This is fine, and I'd just suggest defaulting the runtime guard on and adding tests.  If we think it's dicey by the spec, I think we'd want to either only do it for known buggy user-agents, or have a permanent config option.

Reply:
Thank you. While waiting for confirmation from @PiotrSikora I think I'll make this runtime feature flag enabled by default and add tests for it. Having a permanent config option also sounds plausible for me.

Reply:
I think this change is fine, thanks!

Do you happen to have a tcpdump or debug logs with nghttp2 output that show what's getting transmitted on the wire? I'm wondering if the issue is really a `HEADERS` frame with `END_HEADERS|END_STREAM` flags and empty header block or do we mess something on the wire?

Reply:
@PiotrSikora yes, I just tried using nghttp2.

I think if we have:

```
[  5.053] recv HEADERS frame <length=0, flags=0x05, stream_id=13>
          ; END_STREAM | END_HEADERS
```

Safari throws an error, while: 

```
[  5.050] recv (stream_id=13) grpc-status: 0
[  5.050] recv (stream_id=13) grpc-message: OK
[  5.050] recv HEADERS frame <length=26, flags=0x05, stream_id=13>
          ; END_STREAM | END_HEADERS
          (padlen=0)
[  5.050] send GOAWAY frame <length=8, flags=0x00, stream_id=0>
          (last_stream_id=0, error_code=NO_ERROR(0x00), opaque_data(0)=[])
```

Safari will be fine. But I think what we want is END_STREAM without END_HEADERS with empty HEADERS frame:

```
[  1.362] recv DATA frame <length=0, flags=0x01, stream_id=13>
          ; END_STREAM
```

Complete response when Safari crashes:

```
[  5.032] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=0>
          (window_size_increment=268369921)
[  5.032] send SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  5.049] recv (stream_id=13) :status: 200
[  5.049] recv (stream_id=13) grpc-accept-encoding: identity
[  5.049] recv (stream_id=13) grpc-encoding: identity
[  5.049] recv (stream_id=13) content-type: application/grpc-web-text+proto
[  5.049] recv (stream_id=13) date: Thu, 23 Jul 2020 07:49:27 GMT
[  5.049] recv (stream_id=13) x-envoy-upstream-service-time: 19
[  5.049] recv (stream_id=13) access-control-allow-origin: https://ok.local:8000
[  5.049] recv (stream_id=13) access-control-expose-headers: custom-header-1,grpc-status,grpc-message
[  5.049] recv (stream_id=13) server: envoy
[  5.049] recv HEADERS frame <length=192, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0)
          ; First response header
[  5.049] recv DATA frame <length=28, flags=0x00, stream_id=13>
[  5.053] recv DATA frame <length=52, flags=0x00, stream_id=13>
[  5.053] recv HEADERS frame <length=0, flags=0x05, stream_id=13>
          ; END_STREAM | END_HEADERS
          (padlen=0)
[  5.053] send GOAWAY frame <length=8, flags=0x00, stream_id=0>
          (last_stream_id=0, error_code=NO_ERROR(0x00), opaque_data(0)=[])

```

I have a question though. seems like this patch doesn't work either, we need to somehow set `END_STREAM` when `encodeTrailers` is called. However, I tried to call `nghttp2_submit_data` but it is failing with -510 which is odd since `nghttp2_submit_trailer` is fine with the same session and stream ids. I believe I missed something obvious here.

Reply:
You'll want to make sure that you send an end stream bit somehow.  doing encodeData(empty_buffer, false) should do it.

Reply:
From Filter API perspective, I don't like this approach, I think we should either do:
a) introduce a new return enum to FilterTrailersStatus, to indicate that trailers should not be emit.
b) add `end_stream` flags to `addEncodedData`, end stream there and encodeTrailers return StopIteration

I mildly prefer b) then we have clear H2 frames on wire for gRPC-web without breaking other behavior.

A theoretical case which might break is we have gRPC-web with Safari, though some other H2 client expects empty H2 trailers being proxied from upstream.

Reply:
Interestingly, I also did use the `injectEncodedDataToFilterChain`, but scared of the side-effect, while the response frame seems better:

```
[  5.026] recv (stream_id=13) :status: 200
<SNIP>
[  5.026] recv HEADERS frame <length=191, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0)
          ; First response header
[  5.026] recv DATA frame <length=28, flags=0x00, stream_id=13>
[  5.026] recv DATA frame <length=52, flags=0x01, stream_id=13>
          ; END_STREAM
[  5.026] send GOAWAY frame <length=8, flags=0x00, stream_id=0>
```

```diff
diff --git a/source/extensions/filters/http/grpc_web/grpc_web_filter.cc b/source/extensions/filters/http/grpc_web/grpc_web_filter.cc
index 7595839881..2ad5ba0881 100644
--- a/source/extensions/filters/http/grpc_web/grpc_web_filter.cc
+++ b/source/extensions/filters/http/grpc_web/grpc_web_filter.cc
@@ -222,11 +222,11 @@ Http::FilterTrailersStatus GrpcWebFilter::encodeTrailers(Http::ResponseTrailerMa
   buffer.move(temp);
   if (is_text_response_) {
     Buffer::OwnedImpl encoded(Base64::encode(buffer, buffer.length()));
-    encoder_callbacks_->addEncodedData(encoded, true);
+    encoder_callbacks_->injectEncodedDataToFilterChain(encoded, true);
   } else {
-    encoder_callbacks_->addEncodedData(buffer, true);
+    encoder_callbacks_->injectEncodedDataToFilterChain(buffer, true);
   }
-  return Http::FilterTrailersStatus::Continue;
+  return Http::FilterTrailersStatus::StopIteration;
 }

 void GrpcWebFilter::setupStatTracking(const Http::RequestHeaderMap& headers) {
```

WDYT @lizan?

Reply:
@dio `injectEncodedDataToFilterChain` bypass existing HCM write buffer, so it shouldn't be used. that's why I'm suggesting `addEncodedData`.

Reply:
@lizan updated.

Reply:
I agree this new approach gives more authentic view of what's happening to downstream filters, but I'm wary of the complexity it adds to the already complicated HCM chain, what with some number of filters getting trailers, and a stopiteration which should stop trailers but what if another filter adds trailers back?  Just stripping empty trailers at the codec has the advantages of simplicity, and allowing Envoy-as-an-edge-proxy do fleet wide compliance checks for all upstream servers (we can always add a config option later if someone wants their empty trailers).  If we opt to go this route I think we need a lot more HCM unit tests, verifying if we have filter A-gRPCWeb-B that A and gRPC-web gets trailers, no encodeData(_, true), and B gets encodeData(_, true) no trailers, and then another test where a subsequent downstream filter adds trailers _back_ via addEncodedTrailers and all the right things happen there.
cc @mattklein123

Reply:
@alyssawilk I was feeling adventurous when seeing the data frames seem more correct using this approach tipped by @lizan. At the same time it is scary indeed making changes to HCM this way. I personally prefer to have the simple approach in place first, then since @snowp is making an effort re: filter manager, probably could give clearer path for this case as well.

Reply:
I haven't fully tracked this PR, so apologies, but in general if there is a simple fix I think we should do that. However, I'm a little confused what that fix would be. Are you suggesting to replace empty trailers with an empty data block with end_stream set? If so that seems OK to me.

Either way I agree with @alyssawilk that I would like to see a runtime feature flag for this as well as an integration test.

Reply:
@mattklein123 yes, the original proposal was replacing it by empty data block.

Reply:
@mattklein123 @alyssawilk @lizan is it OK to proceed with the simpler approach first (with runtime guard), then revisit the `addEncodedData`/`addDecodedData` changes after?

Reply:
SGTM!

Reply:
SG. Can you file an issue to track that?

Reply:
On a second thought, since TrailersMap is already a separate class, how about we just add a boolean flag there to indicate it shouldn't be encoded, rather than relying on `empty()`?

Reply:
@lizan are you saying that having empty trailers doesn't automatically infer skipping encoding? And filter needs to explicitly set that?

Reply:
Submitted the issue here: https://github.com/envoyproxy/envoy/issues/12374.

Reply:
> @lizan are you saying that having empty trailers doesn't automatically infer skipping encoding? And filter needs to explicitly set that?

Yes. We can add a flag to https://github.com/envoyproxy/envoy/blob/master/include/envoy/http/header_map.h#L761 ?

Reply:
Cool. Adding flag seems fine. I wanted to confirm the behavior: explicitness of skipping encoding is controlled by filter not the trailers "data".

Reply:
Yep